client/api/src/backend.rs

```diff
@@ -303,58 +302,140 @@ pub trait AuxStore {
 }
 
 /// An `Iterator` that iterates keys in a given block under a prefix.
-pub struct KeyIterator<State, Block> {
+pub struct KeysIter<State, Block>
+where
+	State: StateBackend<HashFor<Block>>,
+	Block: BlockT,
+{
+	inner: <State as StateBackend<HashFor<Block>>>::RawIter,
 	state: State,
-	child_storage: Option<ChildInfo>,
-	prefix: Option<StorageKey>,
-	current_key: Vec<u8>,
-	_phantom: PhantomData<Block>,
+	skip_if_first: Option<StorageKey>,
```

Comment:
Note: the iterators on the storage backend and on `Client` work differently regarding the `start_at`.

For the backend if a `start_at` is specified then it will seek to that key and it will always include that key in the iterator's output. For the `Client` if the key specified in `start_at` matches exactly the first key it encounters then it skips that key.

This inconsistency is a little confusing, but that's how it works.

I've added some extra tests that verify this behavior, and I've also checked that they pass *before* my changes.

Reply:
Yes that's a point to be extremely careful with, retaining the sometime ~ behavior of iteration calls from the existing host functions (all that is touched by `clear_prefix` from sp_externalities).
I remember something ~ with the way error are handled, but I may be confusing with something else.

Reply:
Yeah, I'll probably run a full burn-in before merging this just to make sure nothing's broken.